### PR TITLE
Handle K8s secrets for using private registries

### PR DESCRIPTION
**Pull request by Veebers**
## Description of change

Updates the Juju pod-spec in use so a charm can pass in username/password details used to access a docker image in a private registry (i.e. an image attached to a charm in the charmstore).

This will create any secrets needed and attach them to the resulting K8s pod-spec.

## QA steps

Manually tested using the staging charmstore, microk8s and modified caas mysql charm.
WIP, Needed: Further unit tests.

## Documentation changes

As part of the ongoing work around docker images/caas/resources this change will need to be highlighted (as there are charm changes required to make use of it, although existing caas charms will continue to work).

### CODE REVIEW

File: caas/kubernetes/provider/dockerConfig.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		Auths: map[string]DockerConfigEntry{
+			registryName: dockerEntry,
+		},
+	}
+	return json.Marshal(dockerConfig)
+}
+
+// extractRegistryName returns the registry URL part of an images path
+func extractRegistryURL(imagePath string) (string, error) {
+	// Works for the simplest uses.
```
Comment (wallyworld) [inline caas/kubernetes/provider/dockerConfig.go]: Use this (put the var outside the func up above)

var dockerURLRegexp = regexp.MustCompile(`^(?P<registryURL>[^.]+[.]+[^.]+)\/.*$`)

registryURL := "hub.docker.com"
hasRegistryURL := dockerURLRegexp.MatchString(imagePath)
if hasRegistryURL {
    registryURL = dockerURLRegexp.ReplaceAllString(imagePath, "$registryURL")
}
return registryURL

File: caas/kubernetes/provider/k8s.go [hunk 2]
```diff
@@ -154,6 +154,51 @@ func (k *kubernetesClient) deleteNamespace() error {
 	return errors.Trace(err)
 }
 
+func (k *kubernetesClient) EnsureSecret(appName string, containerSpec *caas.ContainerSpec) error {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: comment
also, pass in ImageDetails, not container

File: caas/kubernetes/provider/k8s.go [hunk 3]
```diff
@@ -154,6 +154,51 @@ func (k *kubernetesClient) deleteNamespace() error {
...
+	if containerSpec.ImageDetails.Password == "" {
+		logger.Debugf("ignoring container, no password provided")
+		return nil
+	}
+	secretData, err := createDockerConfigJSON(&containerSpec.ImageDetails)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	secrets := k.CoreV1().Secrets(k.namespace)
+	secretName := appSecretName(appName, containerSpec.Name)
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: imageSecretName

File: caas/kubernetes/provider/k8s.go [hunk 4]
```diff
@@ -154,6 +154,51 @@ func (k *kubernetesClient) deleteNamespace() error {
...
+func (k *kubernetesClient) deleteSecret(appName, containerName string) error {
+	secretName := appSecretName(appName, containerName)
+	secrets := k.CoreV1().Secrets(k.namespace)
+	err := secrets.Delete(secretName, &v1.DeleteOptions{
+		PropagationPolicy: &defaultPropagationPolicy,
+	})
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: simply return errors.Trace(err)

File: caas/kubernetes/provider/k8s.go [hunk 5]
```diff
@@ -458,11 +503,19 @@ func (k *kubernetesClient) EnsureService(
...
 
-	unitSpec, err := makeUnitSpec(params.PodSpec)
+	unitSpec, err :=
+		makeUnitSpec(appName, params.PodSpec)
 	if err != nil {
 		return errors.Annotatef(err, "parsing unit spec for %s", appName)
 	}
 
+	for _, c := range params.PodSpec.Containers {
+		if err := k.EnsureSecret(appName, &c); err != nil {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: check here that image details has a password and skip if not
then have EnsureSecret return an error if no password

File: caas/kubernetes/provider/k8s.go [hunk 6]
```diff
@@ -1254,3 +1318,9 @@ func deploymentName(appName string) string {
 func resourceNamePrefix(appName string) string {
 	return "juju-" + names.NewApplicationTag(appName).String() + "-"
 }
+
+func appSecretName(appName, containerName string) string {
+	// Possible this will need image name or something else, a pod may have
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: I think container name is unique

File: caas/kubernetes/provider/k8s.go (hunk 2 above)
Comment (Veebers) [reply caas/kubernetes/provider/k8s.go]: I use containerSpec.Name to keep the secret name unique, as a pod may have multiple containers with different images used within (but the same app name).

Comment (wallyworld) [reply caas/kubernetes/provider/k8s.go]: missed that. i think maybe pass in the secret name (constructed outside the method) and image details.
keeps the method a bit tighter in terms of responsibility

File: caas/kubernetes/provider/dockerconfig.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+package provider
+
+import (
+	"encoding/json"
+	"regexp"
+
+	"github.com/juju/juju/caas"
+)
+
+var dockerURLRegexp = regexp.MustCompile(`^(?P<registryURL>([^.]+([.]+[^.\/]+)+))\/.*$`)
```
Comment (wallyworld) [inline caas/kubernetes/provider/dockerconfig.go]: Maybe a comment with an example or 2

File: caas/kubernetes/provider/dockerconfig.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+		Auths: map[string]DockerConfigEntry{
+			registryURL: dockerEntry,
+		},
+	}
+	return json.Marshal(dockerConfig)
+}
+
+// extractRegistryName returns the registry URL part of an images path
+func extractRegistryURL(imagePath string) string {
+	registryURL := "hub.docker.com"
```
Comment (wallyworld) [inline caas/kubernetes/provider/dockerconfig.go]: I just did a manual docker pull to test and this default should be "docker.io"

File: caas/kubernetes/provider/k8s.go [hunk 9]
```diff
@@ -154,6 +154,47 @@ func (k *kubernetesClient) deleteNamespace() error {
 	return errors.Trace(err)
 }
 
+func (k *kubernetesClient) EnsureSecret(imageSecretName, appName string, imageDetails *caas.ImageDetails) error {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: comment

File: caas/kubernetes/provider/k8s.go [hunk 10]
```diff
@@ -458,11 +499,22 @@ func (k *kubernetesClient) EnsureService(
...
 
-	unitSpec, err := makeUnitSpec(params.PodSpec)
+	unitSpec, err :=
+		makeUnitSpec(appName, params.PodSpec)
 	if err != nil {
 		return errors.Annotatef(err, "parsing unit spec for %s", appName)
 	}
 
+	for _, c := range params.PodSpec.Containers {
+		if c.ImageDetails.Password != "" {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: cleaner to do this
if password == "" {
continue
}

less indentation

File: caas/kubernetes/provider/k8s.go [hunk 11]
```diff
@@ -458,11 +499,22 @@ func (k *kubernetesClient) EnsureService(
...
+			imageSecretName := appSecretName(appName, c.Name)
+			if err := k.EnsureSecret(imageSecretName, appName, &c.ImageDetails); err != nil {
+				return errors.Annotatef(err, "creating secrets for container: %s", c.Name)
+			}
+		}
+		cleanups = append(cleanups, func() { k.deleteSecret(appName, c.Name) })
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: and doing the above will fix this bug - we only need cleanup if secret is created

**Comment by Veebers** [comment]
$$merge$$

**Comment by Veebers** [comment]
$$merge$$

**Pull request merged**